File: backend/app/schemas/memory.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.schemas.entity import (
	ConceptCard,
	ItemCard,
	OrganizationCardMemory,
	SceneCardMemory,
	UpdateDynamicInfo,
)
from app.schemas.relation_extract import RelationExtraction


class ParticipantTyped(BaseModel):
	name: str
	type: str


class QueryRequest(BaseModel):
	project_id: int
	participants: Optional[List[str]] = None
	radius: int = 2


class QueryResponse(BaseModel):
	nodes: List[Dict[str, Any]]
	edges: List[Dict[str, Any]]
	fact_summaries: List[str]
	relation_summaries: List[Dict[str, Any]]


class IngestRelationsLLMRequest(BaseModel):
	project_id: int
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None


class IngestRelationsLLMResponse(BaseModel):
	written: int


class ExtractRelationsRequest(BaseModel):
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None


class IngestRelationsFromPreviewRequest(BaseModel):
	project_id: int
	data: RelationExtraction
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None


class IngestRelationsFromPreviewResponse(BaseModel):
	written: int


class ExtractDynamicInfoRequest(BaseModel):
	project_id: int
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	extra_context: Optional[str] = None


class ExtractOnlyRequest(BaseModel):
	project_id: Optional[int] = None
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	extra_context: Optional[str] = None


class UpdateDynamicInfoRequest(BaseModel):
	project_id: int
	data: UpdateDynamicInfo
	queue_size: Optional[int] = 5


class UpdateDynamicInfoResponse(BaseModel):
	success: bool
	updated_card_count: int


class MemoryExtractorInfo(BaseModel):
	code: str
	name: str
	target: str
	preview_supported: bool = True


class MemoryExtractorListResponse(BaseModel):
	items: List[MemoryExtractorInfo]


class ExtractPreviewRequest(BaseModel):
	project_id: Optional[int] = None
	extractor_code: str
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	extra_context: Optional[str] = None
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None
	# 是否自动写入数据库（默认 False，保持原有行为）
	auto_apply: bool = False


class ExtractPreviewResponse(BaseModel):
	extractor_code: str
	preview_data: Dict[str, Any]
	affected_targets: List[Dict[str, Any]] = Field(default_factory=list)
	# 以下字段在 auto_apply=true 时返回
	written: Optional[int] = Field(default=None)
	updated_card_count: Optional[int] = Field(default=None)


class ApplyPreviewRequest(BaseModel):
	project_id: int
	extractor_code: str
	data: Dict[str, Any]
	options: Optional[Dict[str, Any]] = None
	participants: Optional[List[ParticipantTyped]] = None
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None


class ApplyPreviewResponse(BaseModel):
	success: bool
	written: int = 0
	updated_card_count: int = 0
	updated_relation_count: int = 0
	affected_targets: List[Dict[str, Any]] = Field(default_factory=list)
	raw_result: Dict[str, Any] = Field(default_factory=dict)


class ItemStateExtraction(BaseModel):
	items: List[ItemCard] = Field(default_factory=list)


class ConceptStateExtraction(BaseModel):
	concepts: List[ConceptCard] = Field(default_factory=list)


class SceneStateExtraction(BaseModel):
	scenes: List[SceneCardMemory] = Field(default_factory=list)


class OrganizationStateExtraction(BaseModel):
	organizations: List[OrganizationCardMemory] = Field(default_factory=list)


# 一站式记忆提取请求
class ExtractAllRequest(BaseModel):
	project_id: Optional[int] = None
	text: str
	participants: Optional[List[ParticipantTyped]] = None
	llm_config_id: int = 1
	temperature: Optional[float] = None
	max_tokens: Optional[int] = None
	timeout: Optional[float] = None
	extra_context: Optional[str] = None
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None
	# 是否直接写入数据库（默认 False，返回预览）
	auto_apply: bool = False


class TaskResult(BaseModel):
	"""单个提取任务的结果"""
	task: str  # 任务代码
	name: str  # 任务名称
	success: bool
	message: str = ""
	preview_data: Dict[str, Any] = Field(default_factory=dict)
	written: int = 0
	updated_card_count: int = 0
	updated_relation_count: int = 0


class ExtractAllResponse(BaseModel):
	"""一站式记忆提取响应"""
	results: List[TaskResult] = Field(default_factory=list)
	total_written: int = 0
	total_updated_cards: int = 0
	total_updated_relations: int = 0


class ApplyAllRequest(BaseModel):
	"""应用修改后的一站式提取结果"""
	project_id: int
	results: List[TaskResult] = Field(default_factory=list)
	volume_number: Optional[int] = None
	chapter_number: Optional[int] = None

File: frontend/src/renderer/src/api/memory.ts
import http, { aiHttpClient } from './request'
import type { components } from '@renderer/types/generated'

// 使用后端生成的类型（注意部分为 Input/Output 变体）
export type UpdateDynamicInfoOutput = components['schemas']['UpdateDynamicInfo-Output']
export type RelationExtractionOutput = components['schemas']['RelationExtraction-Output']
export interface ParticipantTyped {
	name: string
	type: string
}

export interface MemoryExtractorInfo {
	code: string
	name: string
	target: string
	preview_supported: boolean
}

export interface ExtractPreviewRequest {
	project_id?: number
	extractor_code: string
	text: string
	participants?: ParticipantTyped[]
	llm_config_id: number
	temperature?: number
	max_tokens?: number
	timeout?: number
	extra_context?: string
	volume_number?: number
	chapter_number?: number
	auto_apply?: boolean
}

export interface ExtractPreviewResponse {
	extractor_code: string
	preview_data: Record<string, any>
	affected_targets: Array<Record<string, any>>
}

export interface ApplyPreviewRequest {
	project_id: number
	extractor_code: string
	data: Record<string, any>
	options?: Record<string, any>
	participants?: ParticipantTyped[]
	volume_number?: number
	chapter_number?: number
}

export interface ApplyPreviewResponse {
	success: boolean
	written: number
	updated_card_count: number
	updated_relation_count: number
	affected_targets: Array<Record<string, any>>
	raw_result: Record<string, any>
}

export interface ExtractOnlyRequest {
	project_id?: number
	text: string
	participants?: ParticipantTyped[]
	llm_config_id: number
	temperature?: number
	max_tokens?: number
	timeout?: number
	extra_context?: string
	volume_number?: number
	chapter_number?: number
}

export function extractDynamicInfoOnly(data: ExtractOnlyRequest) {
	return extractMemoryPreview({
		project_id: data.project_id,
		extractor_code: 'character_dynamic',
		text: data.text,
		participants: data.participants,
		llm_config_id: data.llm_config_id,
		temperature: data.temperature,
		max_tokens: data.max_tokens,
		timeout: data.timeout,
		extra_context: data.extra_context,
		volume_number: data.volume_number,
		chapter_number: data.chapter_number,
	}).then(res => res.preview_data as UpdateDynamicInfoOutput)
}

export interface UpdateDynamicInfoOnlyReq {
	project_id: number
	data: UpdateDynamicInfoOutput
	queue_size?: number
}
export type UpdateDynamicInfoOnlyResp = components['schemas']['UpdateDynamicInfoResponse']

export function updateDynamicInfoOnly(data: UpdateDynamicInfoOnlyReq) {
	return applyMemoryPreview({
		project_id: data.project_id,
		extractor_code: 'character_dynamic',
		data: data.data as Record<string, any>,
		options: { queue_size: data.queue_size ?? 5 },
	}).then(res => ({
		success: res.success,
		updated_card_count: res.updated_card_count,
	} as UpdateDynamicInfoOnlyResp))
}

export function listExtractors() {
	return http.get<{ items: MemoryExtractorInfo[] }>('/memory/extractors')
}

export function extractMemoryPreview(data: ExtractPreviewRequest) {
	return http.post<ExtractPreviewResponse>('/memory/extract-preview', data, '/api', { showLoading: false })
}

export function applyMemoryPreview(data: ApplyPreviewRequest) {
	return http.post<ApplyPreviewResponse>('/memory/apply-preview', data, '/api', { showLoading: false })
}

// 入图关系（LLM抽取→Graphiti/Neo4j写入，一步到位）
export type IngestRelationsLLMRequest = components['schemas']['IngestRelationsLLMRequest']
export type IngestRelationsLLMResponse = components['schemas']['IngestRelationsLLMResponse']
export function ingestRelationsLLM(data: IngestRelationsLLMRequest) {
	return http.post<IngestRelationsLLMResponse>('/memory/ingest-relations-llm', data)
}

// 预览→确认入图：使用后端 RelationExtraction-Output
export interface ExtractRelationsOnlyReq {
	text: string
	participants?: ParticipantTyped[]
	llm_config_id: number
	temperature?: number
	max_tokens?: number
	timeout?: number
	volume_number?: number
	chapter_number?: number
}
export function extractRelationsOnly(data: ExtractRelationsOnlyReq) {
	return extractMemoryPreview({
		project_id: undefined,
		extractor_code: 'relation',
		text: data.text,
		participants: data.participants as ParticipantTyped[] | undefined,
		llm_config_id: data.llm_config_id,
		temperature: data.temperature,
		max_tokens: data.max_tokens,
		timeout: data.timeout ?? undefined,
		volume_number: data.volume_number ?? undefined,
		chapter_number: data.chapter_number ?? undefined,
	}).then(res => res.preview_data as RelationExtractionOutput)
}

export type IngestRelationsFromPreviewReq = components['schemas']['IngestRelationsFromPreviewRequest']
export type IngestRelationsFromPreviewResp = components['schemas']['IngestRelationsFromPreviewResponse']
export function ingestRelationsFromPreview(data: IngestRelationsFromPreviewReq) {
	return applyMemoryPreview({
		project_id: data.project_id,
		extractor_code: 'relation',
		data: data.data as Record<string, any>,
		volume_number: data.volume_number ?? undefined,
		chapter_number: data.chapter_number ?? undefined,
	}).then(res => ({ written: res.written } as IngestRelationsFromPreviewResp))
}

// 一站式记忆提取
export interface ExtractAllRequest {
	project_id?: number
	text: string
	participants?: ParticipantTyped[]
	llm_config_id: number
	temperature?: number
	max_tokens?: number
	timeout?: number
	extra_context?: string
	volume_number?: number
	chapter_number?: number
	auto_apply?: boolean
}

export interface TaskResult {
	task: string
	name: string
	success: boolean
	message: string
	preview_data: Record<string, any>
	written: number
	updated_card_count: number
	updated_relation_count: number
}

export interface ExtractAllResponse {
	results: TaskResult[]
	total_written: number
	total_updated_cards: number
	total_updated_relations: number
}

export function extractAll(data: ExtractAllRequest) {
	return aiHttpClient.post<ExtractAllResponse>('/memory/extract-all', data, '/api', { showLoading: false })
}

export function applyAll(data: {
	project_id: number
	results: TaskResult[]
	volume_number?: number
	chapter_number?: number
}) {
	return http.post<ExtractAllResponse>('/memory/apply-all', data, '/api', { showLoading: false })
}
